internal/provider/vaultsecrets/resource_vault_secrets_app_iam_policy.go

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package vaultsecrets
+
+import (
+	"context"
+
+	"github.com/hashicorp/hcp-sdk-go/clients/cloud-resource-manager/stable/2019-12-10/client/resource_service"
+	"github.com/hashicorp/hcp-sdk-go/clients/cloud-resource-manager/stable/2019-12-10/models"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/hashicorp/terraform-provider-hcp/internal/clients"
+	"github.com/hashicorp/terraform-provider-hcp/internal/clients/iampolicy"
+)
+
+// vaultSecretsAppIAMSchema is the schema for the vault secret resource IAM resources
+// (policy/binding). It will be merged with the base policy.
+func vaultSecretsAppIAMSchema(binding bool) schema.Schema {
+	// Determine the description based on if it is for the policy or binding
+	d := "Sets the vault secrets app IAM policy and replaces any existing policy."
+	if binding {
+		d = "Updates the vault secrets app IAM policy to bind a role to a new member. Existing bindings are preserved."
+	}
+
+	return schema.Schema{
+		MarkdownDescription: d,
+		Attributes: map[string]schema.Attribute{
+			"resource_name": schema.StringAttribute{
+				Required:    true,
+				Description: "The app's resource name in the format secrets/project/<project ID>/app/<app ID>.",
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
```

Comment:
Are we not taking `resource_id`? Hmm I was thinking we can take both and let users choose one of them. Using `resource_id` will be good because it is provided in all the other resources and globally unique. My [earlier comment](https://hashicorp.slack.com/archives/C04RWBAC30W/p1712960510269219?thread_ts=1712951592.116149&cid=C04RWBAC30W) was only about ensuring it works with `resource_name`. It did not imply that `resource_id` should not be there.

Reply:
The IAMPolicy APIs require either a resource ID or a resource name but not both. I initially wanted to make both optional for the TFP resource but [GetMutexKey](https://github.com/hashicorp/terraform-provider-hcp/blob/a6a588164a5608633b13413620ed76a89e8bfac6/internal/clients/iampolicy/iam.go#L31) requires a globally unique identifier, that is why I decided to make one of the fields required and drop the other. An option would have been to make both optional and call the resource manager to get the resource ID and use that as the mutex key but the resource manager does not have any API to get a resource by resource name or resource ID

Reply:
> An option would have been to make both optional and call the resource manager to get the resource ID and use that as the mutex key but the resource manager does not have any API to get a resource by resource name or resource ID

That is what I was hoping. Would it be possible to do that? It can be done in subsequent PR.

Reply:
To do that, we will need to create a new API in the resource manager because we don't have an API that can get a resource's ID

Reply:
Based on this thread adding `resource_id` is increasing scope. How about we just add `resource_name` by constructing it in the provider and add the id later?